File: lib/features/sessions/domain/session_repository.dart
import 'package:ten_k_hours/features/sessions/domain/active_session.dart';
import 'package:ten_k_hours/features/sessions/domain/session.dart';

abstract class SessionRepository {
  Future<ActiveSession?> getActive();
  Stream<ActiveSession?> watchActive();

  Future<void> setActive(ActiveSession active);
  Future<void> clearActive();

  Future<Session> insertCompleted({
    required int pursuitId,
    required DateTime startedAt,
    required DateTime endedAt,
    required Duration duration,
  });

  Stream<List<Session>> watchAll(int pursuitId);
  Stream<List<Session>> watchForStats(int pursuitId);

  Future<Duration> totalCountedDurationFor(int pursuitId);

  Stream<Map<DateTime, Duration>> watchDailyTotals(int pursuitId);
}

File: Sources/Models/SessionRepository.swift
import Foundation

enum SessionError: Error, Equatable {
  case alreadyRunning
  case noActiveSession
}

struct StopResult: Equatable {
  let session: Session
  let countedTowardStats: Bool
}

// The slice of persistence `SessionService` needs. Kept minimal so the service
// is unit-testable with an in-memory fake; SwiftData-backed reads for the UI
// (lists, stats, heatmap) go through @Query directly, not this protocol.
protocol SessionRepository {
  func getActive() throws -> ActiveSession?
  func setActive(_ active: ActiveSession) throws
  func clearActive() throws

  func insertCompleted(
    pursuitId: Int,
    startedAt: Date,
    endedAt: Date,
    duration: TimeInterval
  ) throws -> Session
}
